src/archive.cpp

```diff
@@ -86,6 +86,96 @@ namespace zim
     return ret;
   }
 
+  zim::FileImpl::FindxResult findFavicon(FileImpl& impl)
+  {
+    for(auto ns:{'-', 'I'}) {
+      for (auto& path:{"favicon", "favicon.png"}) {
+        auto r = impl.findx(ns, path);
+        if (r.first) {
+          return r;
+        }
+      }
+    }
+    throw EntryNotFound("No favicon found.");
+  }
+
+  unsigned int parseIllustrationPathToSize(const std::string& str) {
+    constexpr auto illuSize = 13; // strlen("Illustration_");
+    auto xpos = str.find('x', illuSize);
+    if (xpos == std::string::npos) {
+      throw std::runtime_error("");
+    }
+    auto atpos = str.find("@1", xpos);
+    if (atpos == std::string::npos) {
+      throw std::runtime_error("");
+    }
+    if (str.substr(illuSize, xpos-illuSize) != str.substr(xpos+1, atpos-xpos-1)) {
+      throw std::runtime_error("");
+    }
+    std::istringstream ss(str.substr(illuSize, xpos-illuSize));
+    unsigned int ret;
+    ss >> ret;
+    return ret;
+  }
+
+  Item Archive::getIllustrationItem(unsigned int size) const {
+    std::ostringstream ss;
+    ss  << "Illustration_" << size << "x" << size << "@" << 1;
+    auto r = m_impl->findx('M', ss.str());
+    if (r.first) {
+      return getEntryByPath(entry_index_type(r.second)).getItem();
+    }
+    // We haven't found the exact entry. Let's "search" for a illustration and
+    // use the first one we found.
+#if 0
+    // We have decided to not implement fallback in case of wrong resolution for now.
+    // We keep this code for reference.
+    r = m_impl->findx('M', "Illustration");
+    auto entry = getEntryByPath(entry_index_type(r.second));
+    if (entry.getPath().find("Illustration") == 0) {
+      return entry.getItem();
+    }
+#endif
+    // For 48x48 illustration, return favicon for older zims.
+    if (size == 48) {
+      auto r = findFavicon(*m_impl);
+      return getEntryByPath(entry_index_type(r.second)).getItem(true);
+    }
+    throw EntryNotFound("Cannot find illustration item.");
+  }
+
+  std::set<unsigned int> Archive::getIllustrationSizes() const {
+    std::set<unsigned int> ret;
+    auto r = m_impl->findx('M', "Illustration_").second;
+    while(true) {
```

Comment:
A `for` loop without a condition will make this code slightly more readable. Using `Archive::findByPath()` that returns an `EntryRange` would be even better at the cost of slight performance loss (which, IMHO, is acceptable here, since `Archive::getIllustrationSizes()` is hardly going to be bottleneck).

Reply:
I've used a for loop.

The problem of `findByPath` is that it searches in `C` namespace only and we are searching in `M`.